src/assign.c

```diff
@@ -819,89 +702,166 @@ const char *memrecycle(SEXP target, SEXP where, int start, int len, SEXP source)
+          if (!temp) {
+            // # nocov start
+            for (int k=0; k<nSourceLevels; ++k) SET_TRUELENGTH(sourceLevelsD[k], 0);
+            savetl_end();
+            error("Unable to allocate working memory of %d bytes to combine factor levels", nAdd*sizeof(SEXP *));
+            // # nocov end
+          }
+          for (int k=0, thisAdd=0; thisAdd<nAdd; ++k) {   // thisAdd<nAdd to stop early when the added ones are all reached
+            SEXP s = sourceLevelsD[k];
+            int tl = TRUELENGTH(s);
+            if (tl) {  // tl negative here
+              if (tl != -nTargetLevels-thisAdd-1) error("Internal error: extra level check sum failed"); // # nocov
+              temp[thisAdd++] = s;
+              SET_TRUELENGTH(s,0);
+            }
+          }
+          savetl_end();
+          setAttrib(target, R_LevelsSymbol, targetLevels=growVector(targetLevels, nTargetLevels + nAdd));
+          for (int k=0; k<nAdd; ++k) {
+            SET_STRING_ELT(targetLevels, nTargetLevels+k, temp[k]);
+          }
+          free(temp);
+        } else {
+          // all source levels were already in target levels, but not with the same integers; we're done
+          savetl_end();
         }
-        free(temp);
-      } else {
-        // all source levels were already in target levels, but not with the same integers; we're done
-        savetl_end();
+        // now continue, but with the mapped integers in the (new) source
+      }
+    }
+  } else if (TYPEOF(target)!=TYPEOF(source)) {
+    // checks up front, otherwise we'd need checks twice in the two branches that cater for 'where' or not
+    bool pun = false;
+    switch(TYPEOF(target)) {
+    case LGLSXP:
+      if (isInteger(source)) {
+        const int *sD = INTEGER(source);
+        for (int i=0; i<slen; ++i) {
+          const int val = sD[i];
+          if (val!=0 && val!=1 && val!=NA_INTEGER) error("Cannot assign %d to a logical column.", val);
+        }
+        pun = true;
+      } else if (isReal(source)) {   // TODO: cater for integer64 here
+        const double *sD = REAL(source);
+        for (int i=0; i<slen; ++i) {
+          const double d = sD[i];
+          if (!ISNAN(d) && d!=0.0 && d!=1.0) error("Cannot assign %f to a logical column.", d);
+        }
+        pun = true;
       }
-      // now continue, but with the mapped integers in the (new) source
+      break;
+    case INTSXP:
+      if (isReal(source)) {
+        int w = INTEGER(isReallyReal(source))[0];  // first fraction present (1-based), 0 if none
+        if (w>0) {
+          warning("Coerced double RHS to integer to match the type of the target column (column %d named '%s'). One or more RHS values contain fractions which have been lost; e.g. item %d with value %f has been truncated to %d.", colnum, colname, w, REAL(source)[w-1], (int)REAL(source)[w-1]);
+        }
+        pun = true;
+      } else if (isLogical(source)) {
+        pun = true;
+      }
+      break;
+    case REALSXP:
```

Comment:
Worth adding punning for `CPLXSXP`? should be similar to `REALSXP`. i can have a go if so.

Reply:
done.